### PR TITLE
DEPR: Deprecate literal json string input to read_json

### LINKED ISSUE #52271: ENH: Deprecate literal json string input to `read_json`
### Feature Type

- [ ] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [X] Removing existing functionality in pandas

### Problem Description

As seen in #29102 (the rejected #29104), and then #46718, determining user intent in user input from `pd.read_json(some_string)` is in general not possible. #46718 is a halfway house in that it explicitly marks some "file extensions" as "you probably wanted to read from a file", but is easily defeated by, for example `pd.read_json("missing.jsonl", lines=True)` (`jsonl` being a common extension for "lines"-formatted json files).

AFAICT, `read_json` is the only `read_XXX` function that accepts a literal representation of the data in its `path_or_buf` argument, so there doesn't seem to be a great deal of precedent here.

### Feature Description

Deprecate literal json input to `pd.read_json`, if one wants to read from a string it should be wrapped in a StringIO.

e.g.
```python
import pandas as pd
from io import StringIO
data = '{"a":{"0":1}}'
# old, proposed for deprecation
df = pd.read_json(data)
# new
df = pd.read_json(StringIO(data))
```

### Alternative Solutions

None

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by rmhowe425**
- [X] closes #52271
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by rmhowe425** [comment]
Both failing tests appear to be failing because the tests pass json literal strings to `read_json`.

**Comment by rmhowe425** [comment]
@mroeschke Would you be able to take a look at this PR and let me know if anything needs updating?

File: doc/source/user_guide/io.rst [hunk 1]
```diff
@@ -2072,6 +2072,10 @@ is ``None``. To explicitly force ``Series`` parsing, pass ``typ=series``
 * ``engine``: Either ``"ujson"``, the built-in JSON parser, or ``"pyarrow"`` which dispatches to pyarrow's ``pyarrow.json.read_json``.
   The ``"pyarrow"`` is only available when ``lines=True``
 
+.. warning::
+
+    Passing json literal strings will be deprecated in a future release of pandas.
```
Comment (mroeschke) [inline doc/source/user_guide/io.rst:2077]: Can you move this to the `read_json` docstring instead?

**Review by mroeschke — CHANGES_REQUESTED** [review]
Could you change the existing tests by wrapping them in `StringIO`?

File: doc/source/user_guide/io.rst (hunk 1 above)
Comment (rmhowe425) [reply doc/source/user_guide/io.rst]: Should be fixed in latest push.

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -268,9 +268,11 @@ Deprecations
 - Deprecated allowing arbitrary ``fill_value`` in :class:`SparseDtype`, in a future version the ``fill_value`` will need to be compatible with the ``dtype.subtype``, either a scalar that can be held by that subtype or ``NaN`` for integer or bool subtypes (:issue:`23124`)
 - Deprecated behavior of :func:`assert_series_equal` and :func:`assert_frame_equal` considering NA-like values (e.g. ``NaN`` vs ``None`` as equivalent) (:issue:`52081`)
 - Deprecated constructing :class:`SparseArray` from scalar data, pass a sequence instead (:issue:`53039`)
+- Deprecated literal json input to :func:`read_json`. Moving forward the method only accepts file-like objects (:issue:`53409`)
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst]: Should be fixed in latest push.

File: doc/source/whatsnew/v2.1.0.rst [hunk 3]
```diff
@@ -268,9 +268,11 @@ Deprecations
...
 - Deprecated positional indexing on :class:`Series` with :meth:`Series.__getitem__` and :meth:`Series.__setitem__`, in a future version ``ser[item]`` will *always* interpret ``item`` as a label, not a position (:issue:`50617`)
 -
 
+
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst]: Should be fixed in latest push.

File: doc/source/user_guide/io.rst [hunk 4]
```diff
@@ -2072,6 +2072,7 @@ is ``None``. To explicitly force ``Series`` parsing, pass ``typ=series``
 * ``engine``: Either ``"ujson"``, the built-in JSON parser, or ``"pyarrow"`` which dispatches to pyarrow's ``pyarrow.json.read_json``.
   The ``"pyarrow"`` is only available when ``lines=True``
 
+
```
Comment (mroeschke) [inline doc/source/user_guide/io.rst:2075]:
```suggestion
```

File: pandas/io/json/_json.py [hunk 5]
```diff
@@ -535,6 +537,11 @@ def read_json(
         By file-like object, we refer to objects with a ``read()`` method,
         such as a file handle (e.g. via builtin ``open`` function)
         or ``StringIO``.
+
+        .. deprecated:: 2.1.0
+            Passing json literal strings will be deprecated in a future release of
```
Comment (mroeschke) [inline pandas/io/json/_json.py:542]:
```suggestion
            Passing json literal strings is deprecated
```

File: pandas/io/json/_json.py [hunk 6]
```diff
@@ -535,6 +537,11 @@ def read_json(
...
+            pandas.
```
Comment (mroeschke) [inline pandas/io/json/_json.py:543]:
```suggestion
```

File: doc/source/user_guide/io.rst (hunk 4 above)
Comment (rmhowe425) [reply doc/source/user_guide/io.rst]: Should be fixed in latest push

File: pandas/io/json/_json.py (hunk 6 above)
Comment (rmhowe425) [reply pandas/io/json/_json.py]: Should be fixed in latest push

**Comment by rmhowe425** [comment]
@mroeschke This PR should be ready for review!

File: pandas/tests/io/json/test_compression.py [hunk 7]
```diff
@@ -23,7 +26,8 @@ def test_compression_roundtrip(compression):
         # explicitly ensure file was compressed.
         with tm.decompress_file(path, compression) as fh:
             result = fh.read().decode("utf8")
-        tm.assert_frame_equal(df, pd.read_json(result))
+            data = StringIO(result)
```
Comment (mroeschke) [inline pandas/tests/io/json/test_compression.py:29]: Can you keep the original indentation here?

File: pandas/tests/io/json/test_deprecated_kwargs.py [hunk 8]
```diff
@@ -10,9 +11,10 @@
 
 def test_good_kwargs():
     df = pd.DataFrame({"A": [2, 4, 6], "B": [3, 6, 9]}, index=[0, 1, 2])
-    with tm.assert_produces_warning(None):
```
Comment (mroeschke) [inline pandas/tests/io/json/test_deprecated_kwargs.py:13]: Can you keep this here?

File: pandas/tests/io/json/test_readlines.py [hunk 9]
```diff
@@ -15,6 +15,20 @@
 from pandas.io.json._json import JsonReader
 
 
+def test_json_deprecation():
```
Comment (mroeschke) [inline pandas/tests/io/json/test_readlines.py:18]:
```suggestion
def test_literal_json_deprecation():
```

Also can you move this test to `test_pandas.py`?

**Comment by rmhowe425** [comment]
@mroeschke This PR should be ready for review!

File: pandas/tests/io/json/test_pandas.py [hunk 10]
```diff
@@ -29,6 +29,20 @@
...
+    # PR 53409
+    expected = DataFrame([[1, 2], [1, 2]], columns=["a", "b"])
+    msg = (
+        "Passing literal json to 'read_json' is deprecated and "
+        "will be removed in a future version. To read from a "
+        "literal string, wrap it in a 'StringIO' object."
+    )
+
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        result = read_json('{"a": 1, "b": 2}\n{"b":2, "a" :1}\n', lines=True)
```
Comment (mroeschke) [inline pandas/tests/io/json/test_pandas.py:42]: Could you test that this also raises a warning with `lines=False`

Comment (rmhowe425) [reply pandas/tests/io/json/test_pandas.py:42]: @mroeschke I added a check in the JsonReader class in `_json.py` to check for the presence of "\n" when lines == False. 

The check seems to be working correctly. However, execution continues after a warning is fired and the code ends up triggering a Value Error: `ValueError: Trailing data`. Seems like the ValueError is overriding the warning and the warning is never caught.

Is there any way to catch the Warning in this scenario?

Comment (rmhowe425) [reply pandas/tests/io/json/test_pandas.py:42]: @mroeschke So just to make sure that I understand, are you asking that add additional tests, or make sure that execution stops once a warning is fired? Or both?

**Comment by rmhowe425** [comment]
@mroeschke PR is ready for review. 

I believe I have addressed your concerns. From what I can tell, the test failures in Numpy Dev are not related to my PR

File: doc/source/whatsnew/v2.1.0.rst [hunk 11]
```diff
@@ -277,6 +277,7 @@ Deprecations
 - Deprecated allowing arbitrary ``fill_value`` in :class:`SparseDtype`, in a future version the ``fill_value`` will need to be compatible with the ``dtype.subtype``, either a scalar that can be held by that subtype or ``NaN`` for integer or bool subtypes (:issue:`23124`)
 - Deprecated behavior of :func:`assert_series_equal` and :func:`assert_frame_equal` considering NA-like values (e.g. ``NaN`` vs ``None`` as equivalent) (:issue:`52081`)
 - Deprecated constructing :class:`SparseArray` from scalar data, pass a sequence instead (:issue:`53039`)
+- Deprecated literal json input to :func:`read_json` (:issue:`53409`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:280]:
```suggestion
- Deprecated literal json input to :func:`read_json`. Wrap literal json string input in `io.StringIO` instead. (:issue:`53409`)
```

**Comment by mroeschke** [comment]
Could you merge in the main branch too?

**Comment by mroeschke** [comment]
Looks like some of the code checks are still failing in the CI

**Comment by rmhowe425** [comment]
@mroeschke Code Checks / Docstring validation passing now. Docbuild and upload test failure doesn't appear to be associated with PR

**Comment by mroeschke** [comment]
You may want to check usage of `read_json` in the the `doc/source` folder for the deprecated behavor. PRs should still pass the doc build

**Comment by rmhowe425** [comment]
@mroeschke Pinging on green!

**Comment by rmhowe425** [comment]
@mroeschke Should be good to go now!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rmhowe425

**Pull request merged**